lib/improver/lapse_rate.py

```diff
@@ -0,0 +1,402 @@
+        self.allbuffers[self.i, :] = buffer
+        self.i += 1
+        return 0.0
+
+
+class LapseRate(object):
+    """Plugin to calculate the lapse rate from orography and temperature
+       cubes.
+
+    Science background:
+
+    References:
+        The method applied here is based on the method used in the 2010 paper:
+        https://rmets.onlinelibrary.wiley.com/doi/abs/10.1002/met.177
+
+    The current lapse rate (LR) FORTRAN code is kept here:
+    http://fcm9/projects/PostProc/browser/PostProc/trunk/
+                                         model_processing/Fss_CalcTLev1Grad.f90
+
+    However the paper doesn't give critical information on certain aspects and
+    they will need more investigation - but for now its important to just have
+    a functional script.
+
+    1) The paper imposes an upper limit to the LR of 0.0098K/m-1
+       (the Dry Adibatic Lapse Rate) but not a lower limit.
+
+       Wheras the current FORTRAN code places a lower limit on the LR of
+       -3.0*DALR in addition to the upper limit.
+
+    2) The paper and FORTRAN code uses a 8x8 neighbourhood, but this was for
+       the older 4km model. More research is required to determine what size
+       neighbourhood is needed for UKV data.
+       In addition - this code will not work with even-numbered neighbourhood
+       sizes. This is because the neighbourhooding code requires a
+       centre point.
+
+    3) Where an invalid lapse rate calculation is returned - should the code
+       return zero or DALR? See "calc_lapse_rate" function.
+
+    Code methodology:
+    1) Apply land/sea mask to temperature and orography datasets. Mask sea
+       points as NaN since image processing module does not recognise Numpy
+       masks.
+    2) Extracts neighbourhoods from both datasets:
+       Apply "generic_filter" image processing module to temperature data
+       to extract a neighbourhood for each single point and save this
+       neighbourhood into a larger array.
+       Repeat for orography data.
+       generic_filter mode="constant" ensures that points beyond edges of
+       dataset will be filled with "cval". (NaN in this case)
+    3) For all the stored orography neighbourhoods - take the neighbours around
+       the central point and create a mask where the height difference from
+       the central point is greater than 35m.
+    4) Loop through array of neighbourhoods and take the height and temperature
+       of all grid points and calculate the
+       temperature/height gradient = lapse rate
+    5) Constrain the lapse rate as > DALR and < -3.0*DALR
+
+    """
+
+    def __init__(self, max_height_diff=35, footprint_len=3,
+                 max_lapse_rate=-3*DALR, min_lapse_rate=DALR):
+        """Initialise class.
+        Args:
+            max_height_diff (float):
+                Maximum allowable height difference between the central point
+                and its neighboours (metres).
+
+            footprint_len (int):
```

Comment:
For consistency with other neighbourhood processing code, could we call this `neighbourhood_size`?